Exercise 1 started. Need to remove words with "--" separators.

LamarMoore opened the pull request:
Exercise 1 is complete and I am submitting for review by a senior developer.

LamarMoore committed:
Exercise 1 started. Need to remove words with "--" separators.

OwenArnold commented on exercises/lamar_moore/ex01_basics/src/main.cpp:
```diff
@@ -0,0 +1,119 @@
+
+	@param p This is the pair which will be swapped around.
+*/
+template<typename A, typename B>
```
Good.

OwenArnold commented on exercises/lamar_moore/ex01_basics/src/main.cpp:
```diff
@@ -0,0 +1,119 @@
+template<typename A, typename B>
+std::pair<B,A> flipPair(const std::pair<A,B> &p)
+{
+    return std::pair<B,A>(p.second, p.first);
```
Could have used std::make_pair here, but otherwise fine.

OwenArnold commented on exercises/lamar_moore/ex01_basics/src/main.cpp:
```diff
@@ -0,0 +1,119 @@
+	for(map<int, string>::const_reverse_iterator iter = sortedList.rbegin(); iter != sortedList.rend(); iter++)
+		outfile<<iter->first<<"\t"<<iter->second<<endl;
+
+	system("notepad.exe HolmesWordUsage.txt");
```
I don't think you would want to leave anything like this in the file.

OwenArnold commented on exercises/lamar_moore/ex01_basics/src/main.cpp:
```diff
@@ -0,0 +1,119 @@
+{
+	bool isPunc = false;
+
+	static const char chars[] = ".,?\'\"!():";
```
Good. Could also have been done with a regex.

OwenArnold commented on exercises/lamar_moore/ex01_basics/src/main.cpp:
```diff
@@ -0,0 +1,119 @@
+
+	@param str String to be examined and if necessary modified.
+*/
+bool checkWord(string &str)
```
Should link declarations and definitions together in the cpp file. Place in an anonymous namespace.

LamarMoore committed:
- CMAKE files added to repository
 - Definitions/declarations merged in cpp file
 - utility functions placed in anonymous namespace

LamarMoore committed:
Exercise 2 complete

OwenArnold commented on exercises/lamar_moore/ex01_basics/src/main.cpp:
```diff
@@ -0,0 +1,149 @@
+
+		static const char chars[] = ".,?\'\"!():";
+
+		for(int i=0; i<strlen(chars); i++)
```
Failing RHEL6 build. strlen is not defined. As the  I think you need to include cstring to use this, or just convert your char \* to as std::string and use std::string::size()

OwenArnold commented on exercises/lamar_moore/ex02_oo_basics/src/Shape.h:
```diff
@@ -0,0 +1,34 @@
+{
+	/** List of possible shape types.
+	*/
+	typedef enum
```
Not a bad solution here. However, 
1. If we are to add more shapes then we'll have to add to this enum. If we're writing a library, this header and enum will be read-only to our clients. So they may not be able to update the definition when they subtype Shape.
2. In a large project (such as Mantid), adding to a header which is used throughout the project could lead to a seriously long recompilation. 
3. This isn't quite as bad as making Shape aware of it's derived types, but it's close. 

For those reasons, I would have chosen to have the type as a std::string.

OwenArnold commented on exercises/lamar_moore/ex02_oo_basics/src/Shape.h:
```diff
@@ -0,0 +1,34 @@
+	class Shape
+	{
+	public:
+		virtual double calculateArea()=0;
```
Good naming. However, all of these methods should be marked const.

OwenArnold commented on exercises/lamar_moore/ex02_oo_basics/src/Shape.h:
```diff
@@ -0,0 +1,34 @@
+		virtual ShapeType getType();
+		virtual int getNumSides();
+	protected:
+		int numSides;
```
I have a preference for enforcing What they need to do, which you have done with your pure virtual functions, but now How. 

The protected `numSides` and `ShapeType` implies that derived classes should set these values. But this is not enforced. int for example does not need to be explicitly initialised, so you the derived classes do not have to set `type` or `numSlides` and the base class portion of the end type will behave incorrectly.

Solve this in either of the following ways.

Have a constructor on the base class which takes both int for numSides and the ShapeType for type as arguments. That way there is no default constructor on the base and derived classes have to call the base class constructor and consciously pass the required arguments. At the same time make both member fields const, so that they cannot be overwritten once set.

Alternately, (my preference) make `getType` and `getNumSides` pure virtual and remove the protected fields so that derived classes are forced to implement everything. I prefer this because then the Shape interface is purely and interface, and not a mix of a base class and and interface.

OwenArnold commented on exercises/lamar_moore/ex02_oo_basics/src/Shape.h:
```diff
@@ -0,0 +1,34 @@
+	protected:
+		int numSides;
+		ShapeType type;
+	};
```
As a minor point, when it comes to coding in mantid, we use _Hungarian notation_ for member fields. We use a prefix of `m_`. This is the only _Hungarian notation_ we do use.

OwenArnold commented on exercises/lamar_moore/ex02_oo_basics/src/Rectangle.cpp:
```diff
@@ -0,0 +1,37 @@
+	@param length Length of the rectangle
+	@param width Width of the rectangle
+	*/
+	Rectangle::Rectangle(double length, double width)
```
Rectangle::Rectangle(double length) : m_length(length)

OwenArnold commented on exercises/lamar_moore/ex02_oo_basics/src/ShapeSorter.cpp:
```diff
@@ -0,0 +1,172 @@
+{
+	/** Functor which is used to compare shape types with a matching value in std::remove_if
+	*/
+	class MatchTypeFunctor
```
Excellent solution!

Two minor points.
1. put these definitions into an anonymous namespace block.
2. passing a shared pointer by value implies change of ownership and can be costly in a tight-loop due to internal increment/decrement of the copies of the shared_ptr. Would be better to pass by `const shared_ptr<Shape>::element_type&` I'd be happy to explain this. But it's not mandatory for you to change this.

OwenArnold commented on exercises/lamar_moore/ex02_oo_basics/src/ShapeSorter.cpp:
```diff
@@ -0,0 +1,172 @@
+	*/
+	void ShapeSorter::printByType(ShapeType type, vector<shared_ptr<Shape>> &shapes)
+	{
+		vector<shared_ptr<Shape>> copy = shapes;
```
I think you could save a lot of allocations deallocations if you used std::copy_if. But I didn't think about that until I looked at the stl.

OwenArnold commented on exercises/lamar_moore/ex02_oo_basics/src/main.cpp:
```diff
@@ -0,0 +1,113 @@
+{
+	/** Obtains a random number between 0 and 3 which corresponds to ShapesTypes.
+	*/
+	ShapeType getRandomShape()
```
Nice way of making any number of shapes.

**Something to consider**
When you're developing in Mantid. We would be using UnitTests CXXTest rather than using hand-rolled main functions for testing. In that case It's nice to see explicit creation rather than the use of random. It also helps a lot with debugging when you want to trace the failure of a particular test case.

LamarMoore committed:
Changes to architecture:
 - use of const for virtual funcitons in Shape inteerface
 - Shape converted to full interface (removed attributes and base class methods. Replaced with interfaces)
 - Implementation class attributes now immutable

OwenArnold commented on exercises/lamar_moore/ex02_oo_basics/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+###############################################################################
+# Executable
+###############################################################################
+
```
Good. Exactly what the CMake list should look like.

LamarMoore committed:
enum ShapeType removed. Shape types are expressed as std:string

LamarMoore committed:
replaced copy.erase with copy_if to reduce allocation/deallocation

OwenArnold commented:
No longer needed.
